mempool/src/pool/collet_txs.rs

```diff
@@ -0,0 +1,247 @@
+    chain::transaction::Transaction,
+    primitives::{Id, Idable},
+};
+use logging::log;
+use utils::{ensure, graph_traversals, shallow_clone::ShallowClone};
+
+/// Transaction entry together with priority
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct EntryByScore<'a> {
+    entry: &'a TxMempoolEntry,
+}
+
+impl PartialOrd for EntryByScore<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl std::ops::Deref for EntryByScore<'_> {
+    type Target = TxMempoolEntry;
+    fn deref(&self) -> &Self::Target {
+        self.entry
+    }
+}
+
+impl Ord for EntryByScore<'_> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.ancestor_score()
+            .cmp(&other.ancestor_score())
+            .then_with(|| self.tx_id().cmp(other.tx_id()))
+    }
+}
+
+impl<'a> From<&'a TxMempoolEntry> for EntryByScore<'a> {
+    fn from(entry: &'a TxMempoolEntry) -> Self {
+        Self { entry }
+    }
+}
+
+pub fn collect_txs<M>(
+    mempool: &Mempool<M>,
+    mut tx_accumulator: Box<dyn TransactionAccumulator>,
+    transaction_ids: Vec<Id<Transaction>>,
+    packing_strategy: PackingStrategy,
+) -> Result<Box<dyn TransactionAccumulator>, BlockConstructionError> {
+    let mempool_tip = mempool.best_block_id();
+    let block_timestamp = tx_accumulator.block_timestamp();
+
+    ensure!(
+        tx_accumulator.expected_tip() == mempool_tip,
+        BlockConstructionError::AccumTipMismatch(tx_accumulator.expected_tip(), mempool_tip),
+    );
+
+    let chainstate = tx_verifier::ChainstateHandle::new(mempool.chainstate_handle.shallow_clone());
+    let chain_config = mempool.chain_config.deref();
+    let utxo_view = tx_verifier::MempoolUtxoView::new(mempool, chainstate.shallow_clone());
+
+    // Transaction verifier to detect cases where mempool is not fully up-to-date with
+    // transaction dependencies.
+    let mut tx_verifier = tx_verifier::create(
+        mempool.chain_config.shallow_clone(),
+        mempool.chainstate_handle.shallow_clone(),
+    );
+
+    let verifier_time = tx_accumulator.block_timestamp();
+
+    let best_index = mempool
+        .blocking_chainstate_handle()
+        .call(|c| c.get_best_block_index())?
+        .expect("best index to exist");
+    let tx_source = TransactionSourceForConnect::for_mempool(&best_index);
+
+    // Use transactions already in the Accumulator to check for uniqueness and to update the
+    // verifier state to update UTXOs they consume / provide.
+    let accum_ids = tx_accumulator
+        .transactions()
+        .iter()
+        .map(|transaction| {
+            let _fee =
+                tx_verifier.connect_transaction(&tx_source, transaction, &verifier_time, None)?;
+            Ok(transaction.transaction().get_id())
+        })
+        .collect::<Result<Vec<_>, TxValidationError>>()?;
+
+    // Set of transactions already placed into the accumulator
+    let mut emitted: BTreeSet<_> = accum_ids.iter().collect();
```

Comment:
I know this was existing code, but since we're here...`tx_accumulater.transactions() is iterated on and then collected for `accum_ids`, and then `accum_ids` is iterated on to collect for `emitted`. Could `emitted` be build within the `accum_ids` `.map())` with a `emmited.push()`?

Reply:
Yea I guess it's possible to do all this in a single pass but the proposed approach with `push` feels somewhat less declarative. Not that it makes a lot of difference, I don't feel strongly either way.

Reply:
All good. This PR is good to go anyway, and we go back to this code then we can update in in a future PR.